In [29]:
import pandas as pd
excel=pd.read_excel("C:/Users/HA/Desktop/financial_data.xlsx")
def excel_to_csv(excel_file_path, csv_file_path):
    # Read the Excel file
    excel_data = pd.read_excel(excel_file_path)
    
    # Save as CSV file
    excel_data.to_csv(csv_file_path, index=False)  # index=False to exclude row numbers from the CSV

    print(f"Successfully converted {excel_file_path} to {csv_file_path}")

# Example usage
excel_to_csv("C:/Users/HA/Desktop/financial_data.xlsx", "C:/Users/HA/Desktop/financial_data.csv")


Successfully converted C:/Users/HA/Desktop/financial_data.xlsx to C:/Users/HA/Desktop/financial_data.csv


In [30]:
df=pd.read_csv("C:/Users/HA/Desktop/financial_data.csv")
df.head(9)

Company  Year  Total Revenues(in millions)  Net Income(in millions)  \
0  Microsoft  2021                       168088                    61271   
1      Tesla  2021                        53823                     5519   
2      Apple  2021                       365817                     9468   
3  Microsoft  2022                       198270                    72738   
4      Tesla  2022                        81462                    12556   
5      Apple  2022                       394328                    99803   
6  Microsoft  2023                       211915                    72361   
7      Tesla  2023                        96773                    14974   
8      Apple  2023                       385285                    96995   

   Total Assets(in millions)  total liabilities(in millions)  \
0                 333779.000                          191791   
1                  62131.000                           30548   
2                 351002.000                          287912   
3                 364840.000                          198298   
4                  82338.000                            3644   
5                 352755.000                          302083   
6                 411976.000                          205753   
7                    106.618                           43009   
8                 352583.000                          290437   

   cash flow from operating activities(in millions)  
0                                             76740  
1                                             11497  
2                                            104038  
3                                             89035  
4                                             14724  
5                                            122151  
6                                             87582  
7                                             13256  
8                                            110543

In [31]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenues(in millions)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income(in millions)'].pct_change() * 100

In [32]:
average_revenue_growth = df.groupby('Company')['Revenue Growth (%)'].mean()

In [33]:
max_net_income_growth = df.groupby('Company')['Net Income Growth (%)'].max()
min_net_income_growth = df.groupby('Company')['Net Income Growth (%)'].min()

print("Max Net Income Growth by Company:")
print(max_net_income_growth)
print("\nMin Net Income Growth by Company:")
print(min_net_income_growth)

Max Net Income Growth by Company:
Company
Apple        954.108576
Microsoft     18.715216
Tesla        127.504983
Name: Net Income Growth (%), dtype: float64

Min Net Income Growth by Company:
Company
Apple        -2.813543
Microsoft    -0.518299
Tesla        19.257725
Name: Net Income Growth (%), dtype: float64


In [34]:
df


Company  Year  Total Revenues(in millions)  Net Income(in millions)  \
0  Microsoft  2021                       168088                    61271   
1      Tesla  2021                        53823                     5519   
2      Apple  2021                       365817                     9468   
3  Microsoft  2022                       198270                    72738   
4      Tesla  2022                        81462                    12556   
5      Apple  2022                       394328                    99803   
6  Microsoft  2023                       211915                    72361   
7      Tesla  2023                        96773                    14974   
8      Apple  2023                       385285                    96995   

   Total Assets(in millions)  total liabilities(in millions)  \
0                 333779.000                          191791   
1                  62131.000                           30548   
2                 351002.000                          287912   
3                 364840.000                          198298   
4                  82338.000                            3644   
5                 352755.000                          302083   
6                 411976.000                          205753   
7                    106.618                           43009   
8                 352583.000                          290437   

   cash flow from operating activities(in millions)  Revenue Growth (%)  \
0                                             76740                 NaN   
1                                             11497                 NaN   
2                                            104038                 NaN   
3                                             89035           17.956071   
4                                             14724           51.351653   
5                                            122151            7.793788   
6                                             87582            6.882030   
7                                             13256           18.795267   
8                                            110543           -2.293269   

   Net Income Growth (%)  
0                    NaN  
1                    NaN  
2                    NaN  
3              18.715216  
4             127.504983  
5             954.108576  
6              -0.518299  
7              19.257725  
8              -2.813543

In [35]:

df['Assets Growth (%)'] = df.groupby('Company')['Total Assets(in millions)'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby('Company')['total liabilities(in millions)'].pct_change() * 100


In [36]:
df['Assets-to-Liabilities Ratio'] = df['Total Assets(in millions)'] / df['total liabilities(in millions)']
df['Cash Flow Growth (%)'] = df.groupby('Company')['cash flow from operating activities(in millions)'].pct_change() * 100

In [37]:
df_summary = df[['Year', 'Company', 'Total Revenues(in millions)', 'Revenue Growth (%)', 'Net Income(in millions)', 'Net Income Growth (%)', 'Assets-to-Liabilities Ratio', 'Cash Flow Growth (%)']]
print(df_summary)

   Year    Company  Total Revenues(in millions)  Revenue Growth (%)  \
0  2021  Microsoft                       168088                 NaN   
1  2021      Tesla                        53823                 NaN   
2  2021      Apple                       365817                 NaN   
3  2022  Microsoft                       198270           17.956071   
4  2022      Tesla                        81462           51.351653   
5  2022      Apple                       394328            7.793788   
6  2023  Microsoft                       211915            6.882030   
7  2023      Tesla                        96773           18.795267   
8  2023      Apple                       385285           -2.293269   

   Net Income(in millions)  Net Income Growth (%)  \
0                    61271                    NaN   
1                     5519                    NaN   
2                     9468                    NaN   
3                    72738              18.715216   
4                    125

In [38]:
df

Company  Year  Total Revenues(in millions)  Net Income(in millions)  \
0  Microsoft  2021                       168088                    61271   
1      Tesla  2021                        53823                     5519   
2      Apple  2021                       365817                     9468   
3  Microsoft  2022                       198270                    72738   
4      Tesla  2022                        81462                    12556   
5      Apple  2022                       394328                    99803   
6  Microsoft  2023                       211915                    72361   
7      Tesla  2023                        96773                    14974   
8      Apple  2023                       385285                    96995   

   Total Assets(in millions)  total liabilities(in millions)  \
0                 333779.000                          191791   
1                  62131.000                           30548   
2                 351002.000                          287912   
3                 364840.000                          198298   
4                  82338.000                            3644   
5                 352755.000                          302083   
6                 411976.000                          205753   
7                    106.618                           43009   
8                 352583.000                          290437   

   cash flow from operating activities(in millions)  Revenue Growth (%)  \
0                                             76740                 NaN   
1                                             11497                 NaN   
2                                            104038                 NaN   
3                                             89035           17.956071   
4                                             14724           51.351653   
5                                            122151            7.793788   
6                                             87582            6.882030   
7                                             13256           18.795267   
8                                            110543           -2.293269   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0                    NaN                NaN                     NaN   
1                    NaN                NaN                     NaN   
2                    NaN                NaN                     NaN   
3              18.715216           9.305858                3.392756   
4             127.504983          32.523217              -88.071232   
5             954.108576           0.499427                4.921990   
6              -0.518299          12.919636                3.759493   
7              19.257725         -99.870512             1080.268935   
8              -2.813543          -0.048759               -3.855232   

   Assets-to-Liabilities Ratio  Cash Flow Growth (%)  
0                     1.740327                   NaN  
1                     2.033881                   NaN  
2                     1.219129                   NaN  
3                     1.839857             16.021631  
4                    22.595499             28.068192  
5                     1.167742             17.409985  
6                     2.002284             -1.631942  
7                     0.002479             -9.970117  
8                     1.213974             -9.502992

1. Revenue Growth (2022-2023)
Microsoft saw a 6.88% revenue growth from 2022 to 2023, showing stable growth.
Tesla had the highest revenue growth in 2022, 51.35%, which drastically slowed down in 2023 to 18.79%.
Apple experienced a slight decline in revenue growth from 2022 to 2023, with -2.29%, indicating potential stagnation in the company's growth trajectory.
2. Net Income Growth (2022-2023)
Tesla had outstanding net income growth in 2022, with a significant 127.5% increase, indicating very strong profitability. In 2023, the net income growth slowed but remained strong at 19.26%.
Microsoft had a positive net income growth of 18.7% in 2022, but the growth turned negative at -0.51% in 2023, showing stagnation in profitability.
Apple witnessed a huge 954% net income growth in 2022, though it slowed to -2.81% in 2023, suggesting some volatility in earnings.
3. Assets Growth
Tesla's assets saw a major jump in 2022 by 32.52%, but then in 2023, assets dropped significantly by -99.87%, possibly due to divestitures or a major restructuring.
Microsoft had consistent asset growth: 9.3% in 2022 and 12.91% in 2023, reflecting a steady expansion of the company's asset base.
Apple showed flat growth with 0.49% in 2022 and -0.04% in 2023, indicating that Apple may not have invested heavily in expanding its asset base.
4. Liabilities Growth
Tesla had an unusual drop in liabilities in 2022 (-88.07%), followed by an extreme increase in 2023 (1080%), which may indicate debt repayment followed by new debt issuance.
Microsoft showed relatively stable liabilities growth in 2022 and 2023 (around 3-4%).
Apple had a slight increase in liabilities in 2022 (4.92%), followed by a slight decrease in 2023 (-3.85%), indicating relatively stable management of debt.
5. Cash Flow from Operating Activities Growth
Tesla had a strong cash flow growth in 2022 (28%), followed by a decline of -9.97% in 2023, indicating possible changes in operational efficiency or investment strategy.
Microsoft's cash flow grew by 16% in 2022 but decreased by -1.63% in 2023, showing some fluctuation in the company's ability to generate cash from its core business.
Apple had positive cash flow growth in 2022 (17.4%), but this also decreased in 2023 by -9.5%.
6. Assets-to-Liabilities Ratio
Tesla's assets-to-liabilities ratio was extremely high in 2022 (22.59), reflecting very low liabilities compared to assets, but dropped significantly in 2023 to 0.0024, which indicates significant changes in the company’s balance sheet.
Microsoft and Apple maintained a stable assets-to-liabilities ratio between 1.16 to 1.84, showing that both companies have more assets than liabilities, which reflects a solid financial standing.
 SO ,Tesla showed extreme volatility in both assets and liabilities, possibly due to debt restructuring or aggressive expansion strategies, but continued to demonstrate strong growth.
Microsoft showed consistent growth in revenues and assets, though its net income growth and cash flow growth slowed in 2023.
Apple had a mixed performance, with declining revenue, net income, and cash flow growth in 2023, but maintained a stable balance sheet with relatively small changes in assets and liabilities.